File: csharp/Tdcv2.Tests/PackPickerScreensTest.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using System.Text.Json;
using Tdcv2.Cli;
using Tdcv2.Packs;
using Xunit;

namespace Tdcv2.Tests;

/// <summary>
/// The pack picker's screens, against the shared fixture.
/// </summary>
/// <remarks>
/// <para>
/// The reference is driven by <c>typescript/scripts/picker-screens.ts</c>, which puts a fake
/// terminal in front of the real picker and records every line it draws after every key. This does
/// the same here, against the same file, so a screen that differs by one space between the two is a
/// failing test rather than something a user notices.
/// </para>
/// <para>
/// One thing this does not exercise, and the other four do: the keys arrive already named. The
/// other implementations decode a terminal's bytes from a stream, so a test can hand them bytes;
/// here <c>ReadKeyName</c> reads the console directly through <c>Console.ReadKey</c> and there is
/// no stream to hand it. What is replayed is everything after that — the navigation and every line
/// of the drawing.
/// </para>
/// </remarks>
public class PackPickerScreensTest
{
    private static readonly string Esc = ((char)27).ToString();

    private static readonly Lazy<JsonDocument> Data = new(() =>
        JsonDocument.Parse(File.ReadAllText(
            Path.Combine(PrngVectorsTest.FixturesDir(), "pack-picker-screens.json"))));

    /// <summary>The screen as the user sees it — clear, home and cursor commands carry no content.</summary>
    private static string[] ToLines(string text)
    {
        foreach (string command in new[] { Esc + "[2J", Esc + "[H", Esc + "[?25l", Esc + "[?25h" })
        {
            text = text.Replace(command, string.Empty, StringComparison.Ordinal);
        }

        if (text.EndsWith('\n'))
        {
            text = text[..^1];
        }

        return text.Split('\n');
    }

    private static List<string> Strings(JsonElement array)
    {
        var out_ = new List<string>();
        foreach (JsonElement item in array.EnumerateArray())
        {
            out_.Add(item.GetString() ?? string.Empty);
        }

        return out_;
    }

    private static List<PackRegistry.Bundle> Bundles()
    {
        var out_ = new List<PackRegistry.Bundle>();
        foreach (JsonElement b in Data.Value.RootElement.GetProperty("bundles").EnumerateArray())
        {
            JsonElement regions = b.GetProperty("regions");
            JsonElement point = b.GetProperty("point");
            out_.Add(new PackRegistry.Bundle(
                b.GetProperty("id").GetString()!,
                b.GetProperty("name").GetString()!,
                b.GetProperty("description").GetString()!,
                b.GetProperty("id").GetString()! + ".zip",
                b.GetProperty("bytes").GetInt64(),
                new string('0', 64),
                null,
                b.GetProperty("locale").GetString(),
                b.GetProperty("country").GetString(),
                Array.Empty<string>(),
                regions.ValueKind == JsonValueKind.Null ? Array.Empty<string>() : Strings(regions),
                point.ValueKind == JsonValueKind.Null
                    ? null
                    : new[] { point[0].GetDouble(), point[1].GetDouble() }));
        }

        return out_;
    }

    /// <summary>One write is one screen, because the picker draws the whole screen at once.</summary>
    private sealed class PerWrite : TextWriter
    {
        public List<string> Written { get; } = new();

        public override Encoding Encoding => Encoding.UTF8;

        public override void Write(string? value) => Written.Add(value ?? string.Empty);
    }

    [Fact]
    public void DrawsWhatTheSharedFixtureSays()
    {
        List<PackRegistry.Bundle> bundles = Bundles();

        foreach (JsonElement run in Data.Value.RootElement.GetProperty("runs").EnumerateArray())
        {
            string name = run.GetProperty("name").GetString()!;
            JsonElement terminal = run.GetProperty("terminal");
            List<string> keys = Strings(run.GetProperty("keys"));
            var installed = new HashSet<string>(Strings(run.GetProperty("installed")));

            int next = 0;
            // Past the end means the script ran out, which is a script that forgot to leave.
            string NextKey() => next < keys.Count ? keys[next++] : "quit";

            var drawn = new PerWrite();
            PackPicker.Decision? decision = PackPicker.RunOn(
                bundles,
                installed,
                drawn,
                NextKey,
                terminal.GetProperty("columns").GetInt32(),
                terminal.GetProperty("rows").GetInt32(),
                terminal.GetProperty("unicode").GetBoolean(),
                terminal.GetProperty("colour").GetBoolean());

            // The first write hides the cursor and carries no screen; everything after it is one
            // draw, ending with the clear on the way out — the empty screen the reference records
            // for the key that left.
            List<string> screens = drawn.Written.GetRange(1, drawn.Written.Count - 1);

            JsonElement want = run.GetProperty("screens");
            Assert.Equal(want.GetArrayLength(), screens.Count);
            for (int i = 0; i < want.GetArrayLength(); i++)
            {
                string after = i == 0 ? "the opening draw" : $"the key \"{keys[i - 1]}\"";
                string expected = string.Join("\n", Strings(want[i]));
                string actual = string.Join("\n", ToLines(screens[i]));
                Assert.True(
                    expected == actual,
                    $"{name}, after {after}:\n--- expected ---\n{expected}\n--- drawn ---\n{actual}");
            }

            JsonElement result = run.GetProperty("result");
            if (result.ValueKind == JsonValueKind.Null)
            {
                Assert.Null(decision);
            }
            else
            {
                Assert.NotNull(decision);
                Assert.Equal(Strings(result.GetProperty("install")), decision!.Install);
                Assert.Equal(Strings(result.GetProperty("remove")), decision.Remove);
            }
        }
    }
}

File: java/src/test/java/io/github/nickliapin/tdc/cli/PackPickerScreensTest.java
package io.github.nickliapin.tdc.cli;

import static org.junit.jupiter.api.Assertions.assertEquals;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.github.nickliapin.tdc.packs.PackRegistry;
import java.io.ByteArrayInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.io.PrintStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Test;

/**
 * The pack picker's screens, against the shared fixture.
 *
 * <p>The reference is driven by {@code typescript/scripts/picker-screens.ts}, which puts a fake
 * terminal in front of the real picker and records every line it draws after every key. This does
 * the same here, against the same file, so a screen that differs by one space between the two is a
 * failing test rather than something a user notices.
 *
 * <p>The seam is {@code PackPicker.Terminal}. A JVM cannot set its own environment variables, so
 * the detection that reads {@code TDCV2_ASCII} and {@code NO_COLOR} cannot be steered from a test
 * the way it can in the other four implementations — the answers are handed in instead, along with
 * the width, the height, and streams that are arrays of bytes rather than a terminal.
 */
class PackPickerScreensTest {

  private static final String ESC = Character.toString(27);
  private static final Path FIXTURE =
      Path.of("..", "fixtures", "cross-language", "pack-picker-screens.json");

  /** What a terminal sends for each key the fixture names; anything else is the character itself. */
  private static String bytesFor(String key) {
    return switch (key) {
      case "up" -> ESC + "[A";
      case "down" -> ESC + "[B";
      case "right" -> ESC + "[C";
      case "left" -> ESC + "[D";
      case "home" -> ESC + "[1~";
      case "end" -> ESC + "[4~";
      case "pageup" -> ESC + "[5~";
      case "pagedown" -> ESC + "[6~";
      case "enter" -> "\r";
      case "space" -> " ";
      case "backspace" -> Character.toString(127);
      case "escape" -> ESC;
      default -> key;
    };
  }

  /** The screen as the user sees it — clear, home and cursor commands carry no content. */
  private static List<String> toLines(String text) {
    String clean = text;
    for (String command : List.of(ESC + "[2J", ESC + "[H", ESC + "[?25l", ESC + "[?25h")) {
      clean = clean.replace(command, "");
    }
    if (clean.endsWith("\n")) {
      clean = clean.substring(0, clean.length() - 1);
    }
    return List.of(clean.split("\n", -1));
  }

  private static JsonNode fixture() {
    try {
      return new ObjectMapper().readTree(Files.readString(FIXTURE));
    } catch (IOException e) {
      throw new UncheckedIOException(e);
    }
  }

  private static List<PackRegistry.Bundle> bundles(JsonNode all) {
    List<PackRegistry.Bundle> out = new ArrayList<>();
    for (JsonNode b : all) {
      JsonNode regions = b.get("regions");
      List<String> in = new ArrayList<>();
      if (!regions.isNull()) {
        for (JsonNode region : regions) {
          in.add(region.asText());
        }
      }
      JsonNode point = b.get("point");
      out.add(
          new PackRegistry.Bundle(
              b.get("id").asText(),
              b.get("name").asText(),
              b.get("description").asText(),
              b.get("id").asText() + ".zip",
              b.get("bytes").asLong(),
              "0".repeat(64),
              null,
              b.get("locale").isNull() ? null : b.get("locale").asText(),
              b.get("country").isNull() ? null : b.get("country").asText(),
              List.of(),
              in,
              point.isNull()
                  ? null
                  : new double[] {point.get(0).asDouble(), point.get(1).asDouble()}));
    }
    return out;
  }

  @Test
  @DisplayName("draws what the shared fixture says, line for line, after every key")
  void drawsWhatTheSharedFixtureSays() {
    JsonNode fixture = fixture();
    List<PackRegistry.Bundle> bundles = bundles(fixture.get("bundles"));

    for (JsonNode run : fixture.get("runs")) {
      String name = run.get("name").asText();
      JsonNode terminal = run.get("terminal");

      StringBuilder keys = new StringBuilder();
      for (JsonNode key : run.get("keys")) {
        keys.append(bytesFor(key.asText()));
      }
      Set<String> installed = new HashSet<>();
      for (JsonNode id : run.get("installed")) {
        installed.add(id.asText());
      }

      // One print is one screen, because the picker draws the whole screen in a single call.
      List<String> drawn = new ArrayList<>();
      PrintStream out =
          new PrintStream(OutputStream.nullOutputStream(), true, StandardCharsets.UTF_8) {
            @Override
            public void print(String text) {
              drawn.add(text);
            }
          };

      PackPicker.Decision decision =
          PackPicker.run(
              bundles,
              installed,
              new PackPicker.Terminal(
                  new ByteArrayInputStream(keys.toString().getBytes(StandardCharsets.UTF_8)),
                  out,
                  terminal.get("columns").asInt(),
                  terminal.get("rows").asInt(),
                  terminal.get("unicode").asBoolean(),
                  terminal.get("colour").asBoolean()));

      // The first print hides the cursor and carries no screen; everything after it is one draw,
      // ending with the clear on the way out — the empty screen the reference records for the key
      // that left.
      List<String> screens = drawn.subList(1, drawn.size());

      JsonNode want = run.get("screens");
      assertEquals(want.size(), screens.size(), "number of screens in " + name);
      for (int i = 0; i < want.size(); i++) {
        List<String> expected = new ArrayList<>();
        for (JsonNode line : want.get(i)) {
          expected.add(line.asText());
        }
        String after = i == 0 ? "the opening draw" : "the key \"" + run.get("keys").get(i - 1) + "\"";
        assertEquals(
            String.join("\n", expected),
            String.join("\n", toLines(screens.get(i))),
            name + ", after " + after);
      }

      JsonNode result = run.get("result");
      if (result.isNull()) {
        assertEquals(null, decision, "result of " + name);
      } else {
        List<String> install = new ArrayList<>();
        for (JsonNode id : result.get("install")) {
          install.add(id.asText());
        }
        List<String> remove = new ArrayList<>();
        for (JsonNode id : result.get("remove")) {
          remove.add(id.asText());
        }
        assertEquals(install, decision.install(), "install of " + name);
        assertEquals(remove, decision.remove(), "remove of " + name);
      }
    }
  }
}
